plugins/modem-manager/fu-mm-device.c

```diff
@@ -1153,6 +1164,106 @@ fu_mm_device_writeln(const gchar *fn, const gchar *buf, GError **error)
 				       error);
 }
 
+gboolean 
+fu_mm_device_write_suspend_ms(FuDevice *device,
+			      const gchar *buf,
+			      GError **error)
+{
+	FuMmDevice *self = FU_MM_DEVICE(device);
+	g_autofree gchar *device_sysfs_path = NULL;
+	g_autofree gchar *autosuspend_delay_filename = NULL;
+
+	/* autosuspend delay updated for a proper firmware update */
+	fu_mm_utils_get_port_info(self->port_mbim, NULL, &device_sysfs_path, NULL, NULL);
+	autosuspend_delay_filename =
+	    g_build_filename(device_sysfs_path, "/power/autosuspend_delay_ms", NULL);
+    
+	if (!fu_mm_device_writeln(autosuspend_delay_filename, buf, error))
+		return FALSE;
+	
+	return TRUE;
+}
+
+guint max(guint x, guint y, guint z)
+{
+	return (x) > (y) ? ((x) > (z) ? (x) : (y)) : ((y) > (z) ? (y):(z));
+}
+
+gboolean
+fu_mm_device_get_device_info(FuDevice *device)
+{
+	FuMmDevice *self = FU_MM_DEVICE(device);
+	g_autofree gchar *device_sysfs_path = NULL;	
+	g_autofree gchar *subsystem_vendor_path = NULL;
+	g_autofree gchar *subsystem_device_path = NULL;	
+	g_autofree gchar *subsystem_vendor = NULL;
+	g_autofree gchar *subsystem_device = NULL;
+	gchar *vendor_id;
+	gchar *device_id;
+
+	const gchar *sdx12_pid[] = {"0xe0da", "0xe0db"};
+	const gchar *sdx55_pid[] = {"0xe0ab", "0xe0b0", "0xe0b1", "0xe0bf", "0xe0c3"};
+	const gchar *sdx62_pid[] = {"0xe0d8", "0xe0d9", "0xe0f5", "0xe0f9"};
+	guint sdx12_lens = sizeof(sdx12_pid)/ sizeof(sdx12_pid[0]);
+	guint sdx55_lens = sizeof(sdx55_pid)/ sizeof(sdx55_pid[0]);
+	guint sdx62_lens = sizeof(sdx62_pid)/ sizeof(sdx62_pid[0]);
+	
+	fu_mm_utils_get_port_info(self->port_mbim, NULL, &device_sysfs_path, NULL, NULL);
+
+	subsystem_vendor_path =
+	    g_build_filename(device_sysfs_path, "/subsystem_vendor", NULL);
+	subsystem_device_path =
+	    g_build_filename(device_sysfs_path, "/subsystem_device", NULL);
+	
+	if (g_file_get_contents (subsystem_vendor_path, &subsystem_vendor, NULL, NULL) &&
+	    g_file_get_contents (subsystem_device_path, &subsystem_device, NULL, NULL))
+	{
+		vendor_id = (gchar *)g_strchomp (subsystem_vendor);
+		device_id = (gchar *)g_strchomp (subsystem_device);
+		
+		for (guint i = 0; i < max(sdx12_lens, sdx55_lens, sdx62_lens); i++)
+		{
+			if (i < sdx12_lens)
+			{
+				for (guint a = 0; a < sdx12_lens; a++)
```

Comment:
did you run ./contrib/setup? the formatting is pretty broken.

Reply:
Hi @hughsie 
Do you mean I need to execute "sudo ./contrib/ci/fwupd_setup_helpers.py install-dependencies -o ubuntu --yes"?

Reply:
just run `./contrib/setup` and it sets everything up.  redo your commits and they get fixed while committing.